### PR TITLE
[graph_trainer] GraphRunner to reduce non-Graph runtime overhead

### PR DESCRIPTION
**Pull request by IvanKobzarev**
Introducing GraphRunner that processes the parameters once and does all extra work only at creation.
Removing this overhead from following runs. 

The overhead is enumerating the parameters of the model on each run. GraphRunner will cache this state inside on creation.

This is especially needed for GradientAccumulation with high number of accumulation N.
As this overhead will be added N times.

GraphRunner is the replacement for run_traced() entry point.
For now keeping both of them.

Depends on #4477.

Agent Description:

## Summary

Add a reusable `GraphRunner` for the non-PP `aot_fx_trace` path. It binds model parameters, model buffers, and trainer-managed graph state once after tracing, then reuses those flattened inputs for every graph invocation.

Previously, every microbatch called `run_traced()`, which repeatedly:

- enumerated module parameters and buffers;
- rebuilt and flattened the model/graph-state pytree;
- unwrapped tensor subclasses; and
- reconstructed the complete graph input list.

`GraphRunner` moves the persistent-state work out of the per-microbatch path. Each invocation now only flattens dynamic user inputs, appends them to the cached state inputs, executes the graph, and reconstructs the outputs. This reduces CPU runner overhead between CUDA graph replays; it does not reduce compilation time or GPU graph execution time.

## Design

`GraphRunner` has three lifecycle phases:

1. Initialization resolves and flattens persistent state tensors once. It records tensor object, metadata, subclass-leaf, and storage identities.
2. Invocation reuses the prepared state inputs and only processes dynamic `(args, kwargs)`.
3. At each training-step boundary, `GraphTrainer` calls `validate_state()` to detect parameter, buffer, or graph-state replacement before stale inputs can be reused.

In-place optimizer and buffer updates remain visible because the runner retains references to the live tensors. Compatible tensor replacement requires constructing a new runner; incompatible metadata or input structure requires retracing.

The logical graph-input order is:

1. module parameters;
2. module buffers;
3. trainer-owned graph-state tensors, currently `parameter.grad` accumulation buffers from #4477;
4. optimizer-state leaves, currently unsupported by `GraphRunner`;
5. runtime `DeviceMesh` inputs used by precompiled SPMD graphs; and
6. user-input leaves from the `(args, kwargs)` pytree.

Tensor subclasses expand into plain-tensor leaves in place within this ordering.

The change also:

- caches the trainable-parameter tuple instead of enumerating parameters for every microbatch;
- records optimizer-state and runtime-mesh input counts in `TracedResult` and precompiled artifacts;
- supports constructing a runner from a loaded precompiled artifact; and
- keeps `run_traced()` as the general-purpose path for tests, debugging, and optimizer-bearing traces.

Pipeline parallelism is intentionally outside this change.

## Performance

The reference benchmark from D118243244 held gradient accumulation fixed at GA=8:

| Runner | Step time |
| --- | ---: |
| Rebuild state on every invocation | 2456.4 ms |
| Bind state once | 2403.1 ms |
| Improvement | 53.3 ms (2.17%) |

The improvement comes from removing repeated Python-side runner work for each microbatch. It is not caused by a change from GA=8 to GA=16.

## Test plan

- 79 focused graph-trainer unit tests passed.
- CUDA graph gradient-accumulation numerical test passed.
- Four-GPU DeepSeek-V3 MinimalAsyncEP completed 10 deterministic training steps with CUDA graphs enabled. The run logged `Recorded CUDA graph`, completed cleanly, and loss decreased from 6.16662 at step 2 to 3.99906 at step 10.
- Changed-file formatting, AST, merge-conflict, large-file, license, flake8, pydoclint, codespell, and link checks passed.

The DSv3 run used PyTorch `2.15.0.dev20260909+cu130` because the installed `2.14.0` build predates the BF16x9 API required on GB300.

### CODE REVIEW

File: torchtitan/experiments/graph_trainer/trainer.py [hunk 1]
```diff
@@ -361,8 +362,23 @@ def _ensure_graph_gradient_state(
                 model,
                 self.optimizers,
             )
+            self._trainable_params = self._graph_gradient_state.parameters
```
Comment (aditvenk) [inline torchtitan/experiments/graph_trainer/trainer.py]: We should not have two sources of truth for populating self._trainable_params; we can call `_get_trainable_parameters` here as well, and inside that, if graph_gradient_state is populated, it can return from there, else construct trainable_params

File: torchtitan/experiments/graph_trainer/trainer.py [hunk 2]
```diff
@@ -445,4 +456,6 @@ def close(self) -> None:
         super().close()
 
         # See Note [explicit cudagraph teardown] in cudagraph.py
```
Comment (aditvenk) [inline torchtitan/experiments/graph_trainer/trainer.py]: nit: comment belongs below the new lines.

File: torchtitan/experiments/graph_trainer/trainer.py [hunk 3]
```diff
@@ -390,10 +406,12 @@ def _validate_inplace_graph_gradient_accumulation_config(self) -> None:
     def _zero_grad(self) -> None:
         if self._graph_gradient_state is None:
             super()._zero_grad()
-            return
-        self._graph_gradient_state.validate_optimizers(self.optimizers)
-        self.optimizers.zero_grad(set_to_none=False)
-        self._graph_gradient_state.validate_bindings()
+        else:
+            self._graph_gradient_state.validate_optimizers(self.optimizers)
```
Comment (aditvenk) [inline torchtitan/experiments/graph_trainer/trainer.py]: It feels like most of this validation can be moved to init time, and perhaps only validate param.grad buffer identity in this hot path. The other stuff really shouldn't change after init?

File: torchtitan/experiments/graph_trainer/precompile.py [hunk 4]
```diff
@@ -294,6 +300,16 @@ def to_traced_result(self, example_inputs: tuple[Any, ...]) -> TracedResult:
             output_subclass_layouts=self.output_subclass_layouts,
             output_spec=self.output_spec,
             state_fqns=self.state_fqns,
+            num_optimizer_state_inputs=getattr(
```
Comment (aditvenk) [inline torchtitan/experiments/graph_trainer/precompile.py]: this backward compatibility feels unnecessary. We could simplify as we don't really expect anyone to be loading precompiled artifacts ocmpiled before this change? (Unless there are test artifacts that I'm missing?)

File: torchtitan/experiments/graph_trainer/runner.py [hunk 5]
```diff
@@ -0,0 +1,294 @@
...
+    Subsequent calls reuse the flattened state inputs to avoid runtime overhead.
+    In-place optimizer and buffer updates remain visible through the retained tensors.
+
+    graph-input order, before tensor-subclass unwrapping:
+
+    1. module parameters, ``named_parameters()`` order
+    2. module buffers, ``named_buffers()`` order
+    3. trainer-owned graph-state tensors,
+        currently ``parameter.grad`` accumulation buffers, ``graph_state`` insertion order
+    4. optimizer-state leaves, which ``GraphRunner`` (not supported yet)
```
Comment (aditvenk) [inline torchtitan/experiments/graph_trainer/runner.py]: Incomplete comment

File: torchtitan/experiments/graph_trainer/runner.py [hunk 6]
```diff
@@ -0,0 +1,294 @@
...
+        self,
+        traced_result: TracedResult,
+        *,
+        module: nn.Module | None = None,
+        graph_state: dict[str, torch.Tensor] | None = None,
+        runtime_meshes: Sequence[DeviceMesh] = (),
+        validate_user_inputs: bool = False,
+        interpreter_cls: type[torch.fx.Interpreter] | None = None,
+    ) -> None:
+        # TODO: Support full training graphs by accepting an optimizer and
```
Comment (aditvenk) [inline torchtitan/experiments/graph_trainer/runner.py]: Do we expect to add this support as part of the landing sprint?

File: torchtitan/experiments/graph_trainer/runner.py [hunk 7]
```diff
@@ -0,0 +1,294 @@
...
+
+    def __init__(
+        self,
+        traced_result: TracedResult,
+        *,
+        module: nn.Module | None = None,
+        graph_state: dict[str, torch.Tensor] | None = None,
+        runtime_meshes: Sequence[DeviceMesh] = (),
+        validate_user_inputs: bool = False,
+        interpreter_cls: type[torch.fx.Interpreter] | None = None,
```
Comment (aditvenk) [inline torchtitan/experiments/graph_trainer/runner.py]: do we need this feature? Or can we add it later when required?

File: torchtitan/experiments/graph_trainer/tests/test_graph_runner.py [hunk 8]
```diff
@@ -0,0 +1,155 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
```
Comment (aditvenk) [inline torchtitan/experiments/graph_trainer/tests/test_graph_runner.py:1]: Acc to agent review there is only situation not tested yet:


GraphRunner has no automated success-path test with DTensor state or explicit runtime meshes. Current tests cover plain tensors and mesh-count rejection only.
A focused test should trace/load a graph containing DTensor model state plus runtime meshes, execute it through GraphRunner, and compare against eager output.

**Review by aditvenk — APPROVED** [review]
LGTM, pl resolve comments before merging.

File: torchtitan/experiments/graph_trainer/trainer.py (hunk 3 above)
Comment (IvanKobzarev) [reply torchtitan/experiments/graph_trainer/trainer.py]: yes, this could be safely simplified.

File: torchtitan/experiments/graph_trainer/precompile.py (hunk 4 above)
Comment (IvanKobzarev) [reply torchtitan/experiments/graph_trainer/precompile.py]: This is mostly for future compatibility, to guard that we are not supporting optimizer state as inputs for now. Technically we do not have this cases now and can remove it.

File: torchtitan/experiments/graph_trainer/runner.py (hunk 6 above)
Comment (IvanKobzarev) [reply torchtitan/experiments/graph_trainer/runner.py]: I think we can do it later, but this should not be hard. But for now agree - we can remove it.

File: torchtitan/experiments/graph_trainer/runner.py (hunk 7 above)
Comment (IvanKobzarev) [reply torchtitan/experiments/graph_trainer/runner.py]: Yes, we can clean it for not

File: torchtitan/experiments/graph_trainer/tests/test_graph_runner.py (hunk 8 above)
Comment (IvanKobzarev) [reply torchtitan/experiments/graph_trainer/tests/test_graph_runner.py:1]: Added the test with it.
